### PR TITLE
added compressed representation

### PR DESCRIPTION
**Pull request by doronz2**
Added compressed representation of the "b" of the BinaryQF form which takes only half of the size of the original element in the form

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -41,7 +49,7 @@ impl BinaryQF {
         let delta = abdelta_triple.delta.clone();
 
         assert_eq!(delta.mod_floor(&BigInt::from(4)), BigInt::one());
-        assert!(delta < BigInt::zero()); // in general delta can be positive but we don't deal with that case
+     //   assert!(delta < BigInt::zero()); // in general delta can be positive but we don't deal with that case
```
Comment (omershlo) [inline src/lib.rs:52]: why is this assertion commented out?

File: src/lib.rs [hunk 2]
```diff
@@ -61,10 +69,81 @@ impl BinaryQF {
     pub fn discriminant(&self) -> BigInt {
         // for negative delta we compute 4ac - b^2
         let abs_delta = BigInt::from(4) * &self.a * &self.c - &self.b * &self.b;
-        assert!(abs_delta > BigInt::zero());
+//        assert!(abs_delta > BigInt::zero());
         -abs_delta
     }
 
+    pub fn compressing_form(&self)->BinaryQF_CompressedForm {
```
Comment (omershlo) [inline src/lib.rs:76]: I think some reference explaining the construction is needed.

File: src/lib.rs [hunk 3]
```diff
@@ -321,6 +400,27 @@ mod tests {
     use crate::curv::arithmetic::traits::Samplable;
     use std::str;
 
+    #[test]
+    fn check_compressed_form(){
```
Comment (omershlo) [inline src/lib.rs:404]: perhaps show that compressed form is indeed 'compressed' compared to decompressed form

File: src/lib.rs [hunk 4]
```diff
@@ -61,10 +69,81 @@ impl BinaryQF {
...
+
+        let new_r = r_square.sqrt() %a.clone();
+        let inv_t = BinaryQF::inv(&comp_b.clone(),&a);
+        let b =(new_r.clone()*inv_t) % a.clone();
+        let c = (&b.pow(2) - &delta) / (BigInt::from(4) * &a);
+        BinaryQF { a, b, c }
+    }
+
+
+    pub fn inv(a: &BigInt, n:&BigInt) -> BigInt
```
Comment (omershlo) [inline src/lib.rs:95]: why not use BigInt::inv(...) here?

File: src/lib.rs (hunk 1 above)
Comment (doronz2) [reply src/lib.rs:52]: This assertion failed during my runs and I saw the comment that " in general delta can be positive", so I thought to comment it out until I figure out how to fix this.

File: src/lib.rs (hunk 2 above)
Comment (doronz2) [reply src/lib.rs:76]: BinaryQF_CompressedForm is a representation of an element in class group.  BinaryQF_CompressedForm is equivalent to BinaryQF except the "b" element which is substituted with a another parameter "b_compressed"  with half the space.  "b_compressed" can be later extracted to the original "b" using the function decompressing_form(BinaryQF_CompressedForm)

File: src/lib.rs (hunk 4 above)
Comment (doronz2) [reply src/lib.rs:95]: I missed that function (I am not familiar with Mpz)...

File: src/lib.rs (hunk 3 above)
Comment (doronz2) [reply src/lib.rs:404]: 👍

Comment (doronz2) [reply src/lib.rs:379]: Added the following assertion: 
`assert!(abd_comp.b_compressed.bit_length()` <= 10+b.bit_length() /2);`
The "+10" bits to this safety as it's not always half the space (in most cases b is around 1000 bits and compressed_b is around 500)

**Pull request closed without merging**